File: workspace/judoku/src/main/java/judoku/solvers/OnlyOnePotentialValueLeftInCell.java
package judoku.solvers;

import java.util.Set;

import judoku.data.Cell;
import judoku.data.Grid;
import judoku.data.OneToNine;

public class OnlyOnePotentialValueLeftInCell implements ISolvingAlgorithm {
	// Does a cell have only one potential value left? Fix it then.
	@Override
	public boolean tryToSolve(Grid grid) {
		for (Cell c: grid.getCells()) {
			if (!c.isFix()) {
				Set<OneToNine> valuesLeft = c.getPotentialValues();

				if (valuesLeft.size() == 1) {
					c.setValue((OneToNine)valuesLeft.toArray()[0]);
				}
				
		// covered by CheckCellsForNoMorePotentialValuesAvailable already?
				else if (valuesLeft.isEmpty()) {
					System.out.println("Error. Grid is: \n" + grid);
					throw new RuntimeException("No values left for cell " + c.id + ". No solution for this grid possible");
				}
		// -- end of ... covered by CheckCellsForNoMorePotentialValuesAvailable already?

			}
		}

		return grid.isSolved();
	}
}

File: workspace/scudoku/src/main/scala/scudoku/solvers/OnlyOnePotentialValueLeftInCell.scala
package scudoku.solvers

/**
 * Java2Scala port from judoku.solvers.OnlyOnePotentialValueLeftInCell
 */

import scudoku.data.Grid
import scudoku.data.Cell

class OnlyOnePotentialValueLeftInCell extends TSolvingAlgorithm {
  // Does a cell have only one potential value left? Fix it then.
  override def tryToSolve(grid: Grid): Boolean = {
    grid.blocks.
      map(block => block.getCells).
      foreach(cells => {
        cells.
        filter(cell => !cell.isFix).
        filter(cell => cell.potentialValues.size == 1).
        map(cell => cell.value_=(cell.potentialValues.head))
      })

    grid.isSolved
  }
}